### PR TITLE
Update Golang to v1.14.4

### PR DESCRIPTION
**Pull request by justaugustus**
**What type of PR is this?**
/kind cleanup
/area dependency release-eng
/sig release

**What this PR does / why we need it**:
Update Golang to v1.14.4

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

/assign @liggitt @cblecker @dims @BenTheElder @listx 
cc: @kubernetes/release-engineering 
ref: https://github.com/kubernetes/release/issues/1216

**Does this PR introduce a user-facing change?**:

```release-note
Update Golang to v1.14.4
```

### CODE REVIEW

**Comment by justaugustus** [comment]
`pull-kubernetes-cross` is an expected failure as the kube-cross image for go1.14.0 doesn't exist yet.
Once this PR passes other tests, I'll build and promote the new kube-cross image.

**Comment by mm4tt** [comment]
FTR, SIG Scalability worked with golang folks on 1.14 release qualification - https://github.com/golang/go/issues/36752
We had some performance regressions but they have been fixed and golang1.14 is good to go from the scalability/performance perspective. 

@justaugustus, for the future, could you add someone from sig-scalability when you update golang to new minor version to double check? This time we managed to beat the regression before first minor release, but for 1.13 a few first patch releases were not good from scale perspective (see https://github.com/golang/go/issues/32828, IIRC 1.13.3 was the first good release). 

Thanks!

**Comment by liggitt** [comment]
looks like etcd tests are unhappy with go1.14:

```
$ go test -race k8s.io/apiserver/pkg/storage/etcd3 |& pp
fatal error: checkptr: unsafe pointer conversion
FAIL	k8s.io/apiserver/pkg/storage/etcd3	0.238s
FAIL
1: running [Created by integration.(*cluster).Launch @ cluster.go:223]
    runtime     panic.go:1112         throw(string(0x2375c38, len=35))
    runtime     checkptr.go:18        checkptrAlignment(Pointer(0xc0002c2130), *_type(0x1))
    bbolt       bucket.go:624         (*Bucket).write(*Bucket(0xc0002bf220), 0x0, 0x0, 0x0)
    bbolt       bucket.go:181         (*Bucket).CreateBucket(*Bucket(0xc000202478), []byte(#1 len=7 cap=7), 0xc0002bf480, 0x1e17936, 0xc000298978)
    bbolt       tx.go:108             (*Tx).CreateBucket(...)
    backend     batch_tx.go:72        (*batchTx).UnsafeCreateBucket(*batchTx(#4), []byte(#1 len=7 cap=7))
    backend     <autogenerated>:1     (*batchTxBuffered).UnsafeCreateBucket(#4, #1, 0x7, 0x7)
    membership  store.go:166          mustCreateBackendBuckets(Backend(0x25e26c0))
    membership  cluster.go:242        (*RaftCluster).SetBackend(...)
    etcdserver  server.go:406         NewServer(ServerConfig(0xc0001d9ca0), 0x0, 0x0, 0x0, 0x0, 0xc00020eb00, 0x1, 0x1, 0xc00020e880, ...)
    integration cluster.go:801        (*member).Launch(*member(#3), 0x0, 0x0)
    integration cluster.go:224        (*cluster).Launch.func1(*cluster(0xc000190900), TB(#3))
    runtime     asm_amd64.s:1373      goexit()
```

xref https://github.com/etcd-io/bbolt/issues/187

cc @enj @jpbetz

**Comment by enj** [comment]
> looks like etcd tests are unhappy with go1.14

If we want to be terrible people, we can do `-gcflags=all=-d=checkptr=0` per the [release notes](https://golang.org/doc/go1.14#compiler) to turn off the new check.  I do not know how that flags interacts with `-race`.

**Comment by jpbetz** [comment]
The improved pointer checking in 1.14 is trying to help us here. We should
trace these down and get etcd to be fully compliant.

**Comment by listx** [comment]
I suppose this PR is blocked on forthcoming(?) etcd fixes...?

**Comment by RainbowMango** [comment]
We'd better take a look at the [mlock issue](https://github.com/golang/go/issues/37436) before update to v1.14.

**Comment by dims** [comment]
Looks like we need to wait on https://github.com/etcd-io/bbolt/pull/201

**Comment by liggitt** [comment]
> We'd better take a look at the [mlock issue](https://github.com/golang/go/issues/37436) before update to v1.14.

go blocking on false positive detection of flawed kernel versions will be resolved in 1.14.1

**Comment by RainbowMango** [comment]
> > We'd better take a look at the [mlock issue](https://github.com/golang/go/issues/37436) before update to v1.14.
> 
> go blocking on false positive detection of flawed kernel versions will be resolved in go1.14.1

Yeah, Go 1.14.1 has been released. We can pick this version I think.

**Comment by justaugustus** [comment]
> > > We'd better take a look at the [mlock issue](https://github.com/golang/go/issues/37436) before update to v1.14.
> > 
> > 
> > go blocking on false positive detection of flawed kernel versions will be resolved in go1.14.1
> 
> Yeah, Go 1.14.1 has been released. We can pick this version I think.

That's the plan!
I was attempting to land [go1.13.9 for Kubernetes v1.18.0](https://github.com/kubernetes/kubernetes/pull/89275), but there are some problematic things I'm running into from the CI perspective ([bazel-related](https://github.com/kubernetes/kubernetes/issues/88553)) that I'd like to try and solve on a go patch bump instead of a minor.

Will report back next week.

**Comment by justaugustus** [comment]
This PR is currently blocked on tests failing due to bazel.
I've provided details here: https://github.com/kubernetes/kubernetes/issues/88553#issuecomment-602411809

**Comment by justaugustus** [comment]
Pending [kube-cross bump](https://github.com/kubernetes/release/pull/1208) & promotion.
I'll also need to rebase out some commits once the [go1.13.9 update](https://github.com/kubernetes/kubernetes/pull/89275) merges.

**Comment by justaugustus** [comment]
> Pending [kube-cross bump](https://github.com/kubernetes/release/pull/1208) & [promotion](https://github.com/kubernetes/k8s.io/pull/687).
> I'll also need to rebase out some commits once the [go1.13.9 update](https://github.com/kubernetes/kubernetes/pull/89275) merges.

Rebased and the kube-cross:v1.14.1-1 image has been promoted, so the next failures we see should hopefully all be legitimate.

**Comment by mm4tt** [comment]
I'll take a look at scalability job failures today

**Comment by justaugustus** [comment]
Thanks @mm4tt! :)

**Comment by mm4tt** [comment]
I took a look at the `pull-kubernetes-e2e-gce-100-performance` and it looks **bad**.  

Test failed because some of the pods timed out. When I dug deeper it turned out that all pods were scheduled on the same node and the logs of the kubelet from that node end in the middle of the test. It suggests that the kubelet got wedged somehow.  

Here are the kubelet logs from that node - https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/88638/pull-kubernetes-e2e-gce-100-performance/1242324434459561984/artifacts/e2e-88638-95a39-minion-group-whdn/kubelet.log. The last entry is from I0324 06:14:15.894303 (while the test was running till ~7:20). 

For comparison here are the logs from some other node - https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/88638/pull-kubernetes-e2e-gce-100-performance/1242324434459561984/artifacts/e2e-88638-95a39-minion-group-8hwt/kubelet.log.

**Comment by mm4tt** [comment]
I also took at look at the kubemark job and it failed because of the same reason (completely different stage of the test, though). 
One hollow (fake) kubelet got wedged.

### Wedged kubelet
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/88638/pull-kubernetes-kubemark-e2e-gce-big/1242324434442784768/artifacts/e2e-88638-ac87c-minion-group-2bfs/kubelet-hollow-node-hxxnt.log
**Last log entries:**
```
I0324 06:22:05.131220       1 kubelet.go:1952] SyncLoop (PLEG): "small-deployment-104-f85b4f675-mzpcz_test-k0ly3l-2(fa09f5f4-404f-4bde-a962-c986de40ff16)", event: &pleg.PodLifecycleEvent{ID:"fa09f5f4-404f-4bde-a962-c986de40ff16", Type:"ContainerStarted", Data:"small-deployment-104-f85b4f675-mzpcz_test-k0ly3l-2_fa09f5f4-404f-4bde-a962-c986de40ff16_0_small-deployment-104_0"}
I0324 06:22:05.131272       1 kubelet.go:1952] SyncLoop (PLEG): "small-deployment-104-f85b4f675-mzpcz_test-k0ly3l-2(fa09f5f4-404f-4bde-a962-c986de40ff16)", event: &pleg.PodLifecycleEvent{ID:"fa09f5f4-404f-4bde-a962-c986de40ff16", Type:"ContainerStarted", Data:"small-deployment-104-f85b4f675-mzpcz_test-k0ly3l-2_fa09f5f4-404f-4bde-a962-c986de40ff16_0"}
E0324 06:22:14.646853       1 container_log_manager.go:216] Failed to stat container log "" after reopen: stat : no such file or directory
E0324 06:22:14.647123       1 container_log_manager.go:216] Failed to stat container log "" after reopen: stat : no such file or directory
E0324 06:22:14.647342       1 container_log_manager.go:216] Failed to stat container log "" after reopen: stat : no such file or directory
E0324 06:22:14.647524       1 container_log_manager.go:216] Failed to stat container log "" after reopen: stat : no such file or directory
E0324 06:22:14.724209       1 container_log_manager.go:216] Failed to stat container log "" after reopen: stat : no such file or directory
E0324 06:22:14.850638       1 container_log_manager.go:216] Failed to stat container log "" after reopen: stat : no such file or directory
E0324 06:22:14.866950       1 container_log_manager.go:216] Failed to stat container log "" after reopen: stat : no such file or directory
```

### Good kubelet
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/88638/pull-kubernetes-kubemark-e2e-gce-big/1242324434442784768/artifacts/e2e-88638-ac87c-minion-group-2bfs/kubelet-hollow-node-hwpvl.log
**Last log entries:**
```
E0324 07:19:23.309247       1 container_log_manager.go:216] Failed to stat container log "" after reopen: stat : no such file or directory
E0324 07:19:33.310360       1 container_log_manager.go:216] Failed to stat container log "" after reopen: stat : no such file or directory
E0324 07:19:33.310930       1 container_log_manager.go:216] Failed to stat container log "" after reopen: stat : no such file or directory
W0324 07:19:41.122870       1 iptables.go:550] Could not set up iptables canary mangle/KUBE-KUBELET-CANARY: error creating chain "KUBE-KUBELET-CANARY": executable file not found in $PATH: 
```


This is serious. I'll reach out to the golang folks to figure out what are the best next steps to debug this.

**Comment by mknyszek** [comment]
@mm4tt If it's truly wedged (which could mean a deadlock somewhere), the best thing here would be to try to get backtraces at the point that it's wedged. Would it be possible to detect this state and send it a `SIGABRT`? That will dump all the goroutine stack traces to STDERR. In order to make sure we get every goroutine's stack (even ones which refuse to be preempted, which could be somewhere deep in the runtime), it would be best to run all kubelets under `GOTRACEBACK=crash`. Is that feasible?

**Comment by mm4tt** [comment]
Thanks, @mknyszek!

It should be doable to run a test with all the changes you requested. I'll try hacking something like this tomorrow.
In the meantime, I'm re-running the tests to get more signals on how often it happens.

/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-e2e-gce-100-performance

**Comment by justaugustus** [comment]
Thanks for the updates. Keep us posted!

**Comment by mm4tt** [comment]
@mknyszek, here is an example wedged kubelet log - https://github.com/mm4tt/kubernetes/blob/golang_1.14.1/kubelet.log
It got wedged around 12:14:02, I killed it with SIGABRT on 12:27:00.  PTAL

**Comment by mknyszek** [comment]
> @mknyszek, here is an example wedged kubelet log - https://github.com/mm4tt/kubernetes/blob/golang_1.14.1/kubelet.log
> It got wedged around 12:14:02, I killed it with SIGABRT at 12:27:00, 10 seconds later it got restarted (you can ignore those logs).
> PTAL

Whew. That took a while to comb through. But I think I have a theory. There's a self-deadlock going on in the timer system.

1. A goroutine does: `timer.Reset -> resettimer -> modtimer -> wakeNetPoller`
    * At this point, the timer its resetting is in timerModifying.
1. Now, at the beginning of `wakeNetPoller`, we get a preemption request (synchronous, so unrelated to asynchronous preemption), so we call into `morestack`.
1. The chain now goes `morestack -> gopreempt_m -> goschedImpl -> schedule -> checkTimers -> runtimer`
1. Now we try to run the timer _we were modifying_, but it's currently being modified, so we loop and `osyield` until it gets out of the modified state.

This can be traced from looking at `goroutine 319` and the very last stack trace produced by a `SIGQUIT` in the log.

CC @ianlancetaylor

I think that theoretically this could have happened prior to go1.14.1 (I don't see why not, anyway). I'm honestly not sure what's special about go1.14.1 here. The only thing I can think of was that racy use of timers is now allowed, and `resettimer` was reimplemented in terms of `modtimer`...

Ah, it might be that the `timerModifying` state is now used to mean more than before? I'm honestly not sure.

Anyway, we should file a bug against Go. It might also be that https://go-review.googlesource.com/c/go/+/224902 helps (but I'm not sure it would fix this).

**Comment by mknyszek** [comment]
Another thought: this can only happen if the timer is in the `timerDeleted` and at the top of the heap, and indeed `(*time.Timer).Reset` does stop (delete) the timer before resetting it. So, this doesn't imply a racy use of timers on Kubernetes' end.

Looking at the state machine in `time.go` again, and the diff between 1.14 and 1.14.1, I don't see how this is a new bug; `resettimer` could have always been preempted (I think?) while a timer was left in the modifying state. With that being said, I think it could _only_ ever happen with the `wakeNetPoller` call, because in `addInitializedTimer` we grab a runtime lock around `cleantimers` and `doaddtimer`, which prevents preemption. 

Maybe what happened is that we accidentally inserted a preemption point? What if it was the case that in `resettimer` before `wakeNetPoller` was actually inlined, but in `addInitializedTimer` it's not. This means we added in a preemption point unintentionally.

One idea for the fix is to just... make sure `wakeNetPoller` gets inlined (and maybe https://go-review.googlesource.com/c/go/+/224902 actually does that), but that seems fragile. We should probably say that while a goroutine owns a timer in `timerModifying` it should not be allowed to be preempted, because it can cause an operation that waits (perhaps the same for other `-ing` states).

Anyway, I'm going to open a Go bug now.

**Comment by liggitt** [comment]
might need to keep an eye on https://github.com/etcd-io/bbolt/issues/187#issuecomment-604214114 as well

**Comment by mm4tt** [comment]
FTR, Golang team fixed the https://github.com/golang/go/issues/38070, the fix will be released some time next week

**Comment by mknyszek** [comment]
> might need to keep an eye on [etcd-io/bbolt#187 (comment)](https://github.com/etcd-io/bbolt/issues/187#issuecomment-604214114) as well

This may or may not be related to https://github.com/golang/go/issues/37688, which is fixed. We'd need more information (and perhaps a reproducer) to understand whether it's the same issue.

**Comment by mm4tt** [comment]
Are we going to try golang1.14.2? Friendly ping

**Comment by ebati** [comment]
go v1.14.2 is released. How can I help to test go v1.14.2?

**Comment by justaugustus** [comment]
> Are we going to try golang1.14.2? Friendly ping

@mm4tt -- Updated! I was waiting for a few of the upstream Go things to settle out.
Will get started on the kube-cross image bump next week.

**Comment by justaugustus** [comment]
kube-cross bump happening here: https://github.com/kubernetes/release/pull/1246

**Comment by justaugustus** [comment]
I bumped to bbolt@v1.3.4, which appears to have fixed the `unsafe` errors in `pull-kubernetes-bazel-test`:
- `//pkg/registry/core/service/storage:go_default_test`
- `//pkg/registry/core/pod/storage:go_default_test`

Testing that one more time:
/test pull-kubernetes-bazel-test

`pull-kubernetes-cross` and `pull-kubernetes-verify` should pass after the new kube-cross image is promoted in https://github.com/kubernetes/k8s.io/pull/765.

Hopefully the last errors are the [ones in `pull-kubernetes-bazel-test` (`//staging/src/k8s.io/apiserver/pkg/apis/config/validation:go_default_test`)](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/88638/pull-kubernetes-bazel-test/1251386911453351937), which seem like quick fixes.

**Comment by liggitt** [comment]
Containers reverted bolt 1.3.4 in https://github.com/containerd/containerd/pull/4156, fyi

**Comment by justaugustus** [comment]
> Containerd reverted bolt 1.3.4 in [containerd/containerd#4156](https://github.com/containerd/containerd/pull/4156), fyi

Eeep! So it sounds like we've got a bit more waiting to do on this one. :/

In the meantime, let's see if there are other issues that fall out, now that the new kube-cross image has been promoted:
/retest

**Comment by BenTheElder** [comment]
eep https://github.com/etcd-io/bbolt/issues/214

**Comment by BenTheElder** [comment]
changes generally LGTM otherwise though .. maybe after the bbolt fix.

**Review by fejta — COMMENTED** [review]
/approve

File: build/root/WORKSPACE
Comment (fejta) [inline build/root/WORKSPACE:7]: FYI ideally we switch to consuming this from repo-infra

Comment (justaugustus) [reply build/root/WORKSPACE]: @fejta -- Thanks. Rebased to pull in your repo-infra@0.0.4 bump.

Comment (fejta) [reply build/root/WORKSPACE]: I'd expect to see this line change to reference 1.14.2?

https://github.com/kubernetes/kubernetes/blob/501781abe4290e19e9b42d292b0b295ea129dea6/build/root/WORKSPACE#L21-L22

Comment (justaugustus) [reply build/root/WORKSPACE]: @fejta -- Ahhhhh, good catch! That got missed in the rebase since the lines moved. Will fix up.

**Review by fejta — APPROVED** [review]
/lgtm
/approve

**Review by fejta — COMMENTED** [review]
/lgtm

**Comment by justaugustus** [comment]
Updated to `kube-cross:v1.14.2-3` (ref: https://github.com/kubernetes/release/pull/1264).

**Comment by justaugustus** [comment]
(Updated kube-cross image to v1.14.2-4 (ref: https://github.com/kubernetes/release/pull/1279, https://github.com/kubernetes/k8s.io/pull/860))

**Comment by justaugustus** [comment]
go1.14.3 is out: https://golang.org/doc/devel/release.html#go1.14
- kube-cross bump: https://github.com/kubernetes/release/pull/1294
- kube-cross promotion: https://github.com/kubernetes/k8s.io/pull/877

**Comment by justaugustus** [comment]
(I think this is going to require a k/repo-infra bump to support the new go version...)

**Comment by justaugustus** [comment]
(k/repo-infra@v0.0.5 bump needs to merge ahead of this: https://github.com/kubernetes/kubernetes/pull/91312)

**Comment by YoyinZyc** [comment]
I noticed there is already a etcd bbolt compatibility fix in progress which should unblock this pr once merged.
https://github.com/etcd-io/bbolt/pull/220

**Comment by justaugustus** [comment]
> I noticed there is already a etcd bbolt compatibility fix in progress which should unblock this pr once merged.
> [etcd-io/bbolt#220](https://github.com/etcd-io/bbolt/pull/220)

@YoyinZyc -- thanks for the heads up! I've been eagerly watching that PR :)

**Comment by Huang-Wei** [comment]
@justaugustus Are we going to adopt go 1.14 in k8s 1.19 release?

**Comment by justaugustus** [comment]
@Huang-Wei -- that's the plan; we need the fix in bbolt before we can move forward though.

I'm planning to work on updating this PR (next week) for go1.14.4 as well, now that that's available.

/pull-kubernetes-e2e-gce-ubuntu-containerd

**Comment by justaugustus** [comment]
go1.14.4 is out: https://golang.org/doc/devel/release.html#go1.14
- kube-cross bump: https://github.com/kubernetes/release/pull/1339
- kube-cross promotion: https://github.com/kubernetes/k8s.io/pull/944

(I think this is going to require a k/repo-infra bump to support the new go version...)

**Comment by BenTheElder** [comment]
Still blocked on etcd/bbolt?

**Comment by justaugustus** [comment]
> Still blocked on etcd/bbolt?

@BenTheElder -- Hopefully not anymore! :)
I've opened https://github.com/kubernetes/kubernetes/pull/92350 to bump bbolt to the newest release, v1.3.5 (ref: https://github.com/etcd-io/bbolt/issues/214#issuecomment-645574091).

I'll get this rebased once that merges AND we update k/repo-infra.

**Comment by justaugustus** [comment]
`pull-kubernetes-e2e-kind` has been flaking and `pull-kubernetes-verify` failed in 2f6d8e3 due to a whitespace issue in `build/root/WORKSPACE`, but everything else is passing with the test commits added.

**Comment by justaugustus** [comment]
F I N A L L Y .
/honk
/pony party

**Comment by RainbowMango** [comment]
Congrats! It's quite a long ride!

**Comment by BenTheElder** [comment]
reminder that the IPV6 job is not required for merge 🙃

**Pull request merged**
